lib/parsers/yarn-lock-parse-base.ts

```diff
@@ -0,0 +1,190 @@
+import * as pMap from 'p-map';
+
+import {
+  LockfileParser,
+  PkgTree,
+  DepTreeDep,
+  Dep,
+  ManifestFile,
+  getTopLevelDeps,
+  Lockfile,
+  LockfileType,
+  createDepTreeDepFromDep,
+} from './';
+import { eventLoopSpinner } from 'event-loop-spinner';
+import {
+  InvalidUserInputError,
+  OutOfSyncError,
+  TreeSizeLimitError,
+} from '../errors';
+import { config } from '../config';
+
+const EVENT_PROCESSING_CONCURRENCY = 5;
+
+export type YarnLockFileTypes = LockfileType.yarn;
+
+export interface YarnLockDeps {
+  [depName: string]: YarnLockDep;
+}
+
+export interface YarnLockBase<T extends YarnLockFileTypes> {
+  type: string;
+  object: YarnLockDeps;
+  dependencies?: YarnLockDeps;
+  lockfileType: T;
+}
+
+export interface YarnLockDep {
+  version: string;
+  dependencies?: {
+    [depName: string]: string;
+  };
+  optionalDependencies?: {
+    [depName: string]: string;
+  };
+}
+
+export abstract class YarnLockParseBase<T extends YarnLockFileTypes>
+  implements LockfileParser {
+  private treeSize: number;
+  private eventLoopSpinRate = 20;
+
+  constructor(private type: T) {
+    // Number of dependencies including root one.
+    this.treeSize = 1;
+  }
+
+  public abstract parseLockFile(lockFileContents: string): Lockfile;
+
+  public async getDependencyTree(
+    manifestFile: ManifestFile,
+    lockfile: Lockfile,
+    includeDev = false,
+    strict = true,
+  ): Promise<PkgTree> {
+    if (lockfile.type !== this.type) {
+      throw new InvalidUserInputError(
+        'Unsupported lockfile provided. ' +
+          'Please provide `package-lock.json`.',
+      );
+    }
+    const yarnLock = lockfile as YarnLockBase<T>;
+
+    const depTree: PkgTree = {
+      dependencies: {},
+      hasDevDependencies: !_.isEmpty(manifestFile.devDependencies),
+      name: manifestFile.name,
+      size: 1,
+      version: manifestFile.version || '',
+    };
+
+    const nodeVersion = _.get(manifestFile, 'engines.node');
+    if (nodeVersion) {
+      _.set(depTree, 'meta.nodeVersion', nodeVersion);
+    }
+
+    const topLevelDeps: Dep[] = getTopLevelDeps(manifestFile, includeDev);
+    // asked to process empty deps
+    if (_.isEmpty(manifestFile.dependencies) && !includeDev) {
+      return depTree;
+    }
+
+    await pMap(
+      topLevelDeps,
+      (dep) => this.resolveDep(dep, depTree, yarnLock, strict),
+      { concurrency: EVENT_PROCESSING_CONCURRENCY },
+    );
+
+    depTree.size = this.treeSize;
+    return depTree;
+  }
+
+  private async buildSubTree(
+    lockFile: YarnLockBase<T>,
+    tree: DepTreeDep,
+    strict: boolean,
+  ): Promise<DepTreeDep> {
+    const queue = [{ path: [] as string[], tree }];
+    while (queue.length > 0) {
+      const queueItem = queue.pop()!;
+      const depKey = `${queueItem.tree.name}@${queueItem.tree.version}`;
+      const dependency = lockFile.object[depKey];
+      if (!dependency) {
+        if (strict) {
+          throw new OutOfSyncError(queueItem.tree.name!, 'yarn');
```

Comment:
Good catch @gitphill 

```suggestion
          throw new OutOfSyncError(queueItem.tree.name!, this.type);
```
@dkontorovskyy Notice you will need to change the `out-of-sync-error` to except type YarnLockFileTypes as parameter